File: src/main/java/com/inqbarna/iqlocation/LocationPermissionRequestDelegate.java
package com.inqbarna.iqlocation;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.PendingIntent;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResult;
import com.google.android.gms.location.LocationSettingsStates;

import java.util.ArrayList;
import java.util.List;

public class LocationPermissionRequestDelegate extends PermissionRequestDelegate {

    public static final int  DEFAULT_RC_RESOLVE_ERROR    = 1;
    public static final int  DEFAULT_RC_RESOLVE_SETTINGS = 2;

    private int mRequestCodeResolveError       = DEFAULT_RC_RESOLVE_ERROR;
    private int mRequestCodeResolveSettings    = DEFAULT_RC_RESOLVE_SETTINGS;

    private final PermissionDelegateCallbacks mCallbacks;
    private       GoogleApiClient             mApiClient;

    private final GoogleApiClient.ConnectionCallbacks mClientCallback = new GoogleApiClient.ConnectionCallbacks() {
        @Override
        public void onConnected(@Nullable Bundle bundle) {
            if (null != mOnConnected) {
                mOnConnected.run();
                mOnConnected = null;
            }
        }

        @Override
        public void onConnectionSuspended(int i) {
        }
    };

    private boolean mResolvingError;
    private final GoogleApiClient.OnConnectionFailedListener mConnectionFailedCallback = new GoogleApiClient.OnConnectionFailedListener() {
        @Override
        public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
            if (mResolvingError) {
                // Already attempting to resolve an error.
                return;
            } else if (connectionResult.hasResolution()) {
                try {
                    doResolveResolvable(connectionResult.getResolution(), mRequestCodeResolveError);
                } catch (IntentSender.SendIntentException e) {
                    // There was an error with the resolution intent. Try again.
                    mApiClient.connect();
                }
            } else {
                // Show dialog using GoogleApiAvailability.getErrorDialog()
                final String errorString = GoogleApiAvailability.getInstance().getErrorString(connectionResult.getErrorCode());
                Log.e(TAG, "Error connecting to GoogleApiClient: " + errorString);
                failIfRequestedLocation();
                mResolvingError = true;
            }
        }
    };

    public void setRequestCodeResolveError(int requestCodeResolveError) {
        this.mRequestCodeResolveError = requestCodeResolveError;
    }

    public void setRequestCodeResolveSettings(int requestCodeResolveSettings) {
        this.mRequestCodeResolveSettings = requestCodeResolveSettings;
    }

    protected void doResolveResolvable(PendingIntent resolution, int requestCode) throws IntentSender.SendIntentException {
        if (mResolvingError) {
            return;
        }
        mResolvingError = true;
        final IntentSender intentSender = resolution.getIntentSender();
        if (null != getFragment()) {
            getFragment().startIntentSenderForResult(intentSender, requestCode, null, 0, 0, 0, null);
        } else {
            getContext().startIntentSenderForResult(intentSender, requestCode, null, 0, 0, 0);
        }
    }

    private void failIfRequestedLocation() {
        if (null != mOnConnected) {
            mCallbacks.onPermissionDenied();
            mOnConnected = null;
        }
    }

    @Override
    protected Options getOptions() {
        return (Options)super.getOptions();
    }

    private Runnable mOnConnected;

    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == mRequestCodeResolveError) {
            mResolvingError = false;
            if (resultCode == Activity.RESULT_OK) {
                if (!mApiClient.isConnected() && !mApiClient.isConnecting()) {
                    mApiClient.connect();
                }
            }
            return true;
        } else if (requestCode == mRequestCodeResolveSettings) {
            mResolvingError = false;
            if (resultCode == Activity.RESULT_OK) {
                processLocationSettings(LocationSettingsStates.fromIntent(data));
            } else {
                mCallbacks.onPermissionDenied();
            }
            return true;
        } else {
            return false;
        }
    }

    public static class Options extends PermissionRequestDelegate.Options {
        final boolean checkSettings;
        final List<LocationRequest> mRequests;

        public Options(Builder builder) {
            super(builder);
            checkSettings = builder.checkSettings;
            mRequests = builder.locationRequests;
        }

        public static class Builder extends PermissionRequestDelegate.Options.Builder<Builder> {
            boolean checkSettings = false;
            List<LocationRequest> locationRequests = new ArrayList<>();

            public Builder satisfyRequests(@NonNull List<? extends LocationRequest> requests) {
                this.checkSettings = true;
                locationRequests.addAll(requests);
                return this;
            }
            
            public Options build() {
                return new Options(this);
            }
        }
    }

    @Override
    public void onSaveState(Bundle state) {
        super.onSaveState(state);
    }

    private PermissionDelegateCallbacks mBridgedCallbacks = new PermissionDelegateCallbacks() {
        @Override
        public void showRequestPermissionsDialog(DialogInterface.OnClickListener accept, DialogInterface.OnClickListener deny) {
            mCallbacks.showRequestPermissionsDialog(accept, deny);
        }

        @Override
        public void onPermissionGranted() {
            if (!getOptions().checkSettings) {
                mCallbacks.onPermissionGranted();
            } else {
                beginSettingsCheck();
            }
        }

        @Override
        public void onPermissionDenied() {
            mCallbacks.onPermissionDenied();
        }
    };

    private static final Options DEFAULT_OPTIONS = new Options.Builder().build();

    public LocationPermissionRequestDelegate(Activity context, PermissionDelegateCallbacks callbacks, @Nullable Bundle state) {
        this(context, callbacks, DEFAULT_OPTIONS, state);
    }

    public LocationPermissionRequestDelegate(Activity context, PermissionDelegateCallbacks callbacks, @NonNull Options options, @Nullable Bundle state) {
        this(context, callbacks, null, options, state);
    }

    public LocationPermissionRequestDelegate(Fragment fragment, PermissionDelegateCallbacks callbacks, @Nullable Bundle state) {
        this(fragment, callbacks, DEFAULT_OPTIONS, state);
    }

    public LocationPermissionRequestDelegate(Fragment fragment, PermissionDelegateCallbacks callbacks, @NonNull Options options, @Nullable Bundle state) {
        this(fragment.getActivity(), callbacks, fragment, options, state);
    }

    public LocationPermissionRequestDelegate(Activity context, PermissionDelegateCallbacks callbacks, Fragment fragment, @NonNull Options options, @Nullable Bundle state) {
        super(context, null, fragment, options, state);
        setCallbacks(mBridgedCallbacks);
        mCallbacks = callbacks;

        if (options.checkSettings) {
            mApiClient = new GoogleApiClient.Builder(getContext(), mClientCallback, mConnectionFailedCallback)
                    .addApi(LocationServices.API)
                    .build();
        }
    }


    public final void checkLocationPermission() {
        checkPermissionsAll(new String[] { Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION});
    }

    private void beginSettingsCheck() {
        Runnable task = new Runnable() {
            @Override
            public void run() {
                LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder();
                builder.addAllLocationRequests(getOptions().mRequests);

                LocationServices.SettingsApi.checkLocationSettings(mApiClient, builder.build())
                        .setResultCallback(
                                new ResultCallback<LocationSettingsResult>() {
                                    @Override
                                    public void onResult(@NonNull LocationSettingsResult locationSettingsResult) {
                                        final Status status = locationSettingsResult.getStatus();
                                        if (!status.isSuccess()) {
                                            if (status.hasResolution()) {
                                                final PendingIntent resolution = locationSettingsResult.getStatus().getResolution();
                                                try {
                                                    doResolveResolvable(resolution, mRequestCodeResolveSettings);
                                                } catch (IntentSender.SendIntentException e) {
                                                    Log.e(TAG, "Cannot resolve error", e);
                                                    mCallbacks.onPermissionDenied();
                                                }
                                            } else {
                                                mCallbacks.onPermissionDenied();
                                            }
                                        } else {
                                            processLocationSettings(locationSettingsResult.getLocationSettingsStates());
                                        }
                                    }
                                }
                        );
            }
        };

        if (mApiClient.isConnected()) {
            task.run();
        } else if (mApiClient.isConnecting()) {
            mOnConnected = task;
        } else {
            mCallbacks.onPermissionDenied();
        }
    }

    protected void processLocationSettings(LocationSettingsStates locationSettingsStates) {
        if (locationSettingsStates.isLocationPresent() && locationSettingsStates.isLocationUsable()) {
            mCallbacks.onPermissionGranted();
        } else {
            mCallbacks.onPermissionDenied();
        }
    }

    public void onStart() {
        if (null != mApiClient) {
            mApiClient.connect();
        }
    }

    public void onStop() {
        if (null != mApiClient) {
            mApiClient.disconnect();
        }
    }

    public static AlertDialog defaultLocationDialog(Context context, DialogInterface.OnClickListener accept, DialogInterface.OnClickListener decline) {
        final Resources resources = context.getResources();
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(context);

        // set title
        alertDialogBuilder.setTitle(resources.getString(R.string.request_location_title));

        // set dialog message
        alertDialogBuilder.setMessage(resources.getString(R.string.request_location_msg)).setCancelable(false);

        alertDialogBuilder.setPositiveButton(resources.getString(android.R.string.ok), accept).setNegativeButton(resources.getString(android.R.string.cancel), decline);

        // create alert dialog
        return alertDialogBuilder.create();
    }
}

File: library/src/main/java/com/inqbarna/iqlocation/LocationPermissionRequestDelegate.kt
/*
 * Copyright 2021 InQBarna Kenkyuu Jo SL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.inqbarna.iqlocation

import android.Manifest
import android.app.Activity
import android.app.PendingIntent
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentSender.SendIntentException
import android.os.Bundle
import androidx.fragment.app.Fragment
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.LocationSettingsRequest
import com.google.android.gms.location.LocationSettingsStates
import com.google.android.gms.location.LocationSettingsStatusCodes
import com.google.android.gms.location.SettingsClient
import org.slf4j.LoggerFactory
import java.util.*

open class LocationPermissionRequestDelegate private constructor(context: Activity,
                                        callbacks: PermissionDelegateCallbacks,
                                        fragment: Fragment?,
                                        options: Options,
                                        state: Bundle?) : PermissionRequestDelegate(context, null, fragment, options, state) {
    var requestCodeResolveSettings = DEFAULT_RC_RESOLVE_SETTINGS
    private val settingsClient: SettingsClient?
    private lateinit var interceptedCallbacks: PermissionDelegateCallbacks


    private var resolvingError = false

    private val upstreamCallbacks: PermissionDelegateCallbacks = object : PermissionDelegateCallbacks {
        override fun showRequestPermissionsDialog(accept: DialogInterface.OnClickListener, deny: DialogInterface.OnClickListener) {
            interceptedCallbacks.showRequestPermissionsDialog(accept, deny)
        }

        override fun onPermissionGranted(alreadyGranted: Boolean) {
            if (!getOptions().checkSettings) {
                interceptedCallbacks.onPermissionGranted(alreadyGranted)
            } else {
                beginSettingsCheck(alreadyGranted)
            }
        }

        override fun onPermissionDenied() {
            interceptedCallbacks.onPermissionDenied()
        }
    }

    init {
        this.callbacks = upstreamCallbacks
        interceptedCallbacks = callbacks
        settingsClient = if (options.checkSettings) {
            LocationServices.getSettingsClient(context)
        } else {
            null
        }
    }

    @JvmOverloads
    constructor(context: Activity, callbacks: PermissionDelegateCallbacks, state: Bundle? = null, options: Options = defaultOptions()) : this(context, callbacks, null, options, state)
    @JvmOverloads
    constructor(fragment: Fragment, callbacks: PermissionDelegateCallbacks, state: Bundle? = null, options: Options = defaultOptions()) : this(fragment.requireActivity(), callbacks, fragment, options, state)

    @Throws(SendIntentException::class)
    protected fun doResolveResolvable(resolution: PendingIntent, requestCode: Int) {
        if (resolvingError) {
            return
        }
        resolvingError = true
        val intentSender = resolution.intentSender
        if (null != fragment) {
            fragment.startIntentSenderForResult(intentSender, requestCode, null, 0, 0, 0, null)
        } else {
            context.startIntentSenderForResult(intentSender, requestCode, null, 0, 0, 0)
        }
    }

    private fun getOptions(): Options {
        return options as Options
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        return if (requestCode == requestCodeResolveSettings) {
            resolvingError = false
            if (resultCode == Activity.RESULT_OK && data != null) {
                processLocationSettings(LocationSettingsStates.fromIntent(data), false)
            } else {
                interceptedCallbacks.onPermissionDenied()
            }
            true
        } else {
            false
        }
    }

    class Options private constructor(
            internal val checkSettings: Boolean,
            internal val requests: List<LocationRequest>,
            baseOptions: PermissionRequestDelegate.Options
    ) : PermissionRequestDelegate.Options(baseOptions) {

        class Builder {
            private var checkSettings = false
            private var locationRequests: MutableList<LocationRequest> = ArrayList()

            private val baseBuilder = PermissionRequestDelegate.Options.Builder()
            fun satisfyRequests(requests: List<LocationRequest>): Builder = apply {
                checkSettings = true
                locationRequests.addAll(requests)
            }

            fun satisfyRequest(request: LocationRequest): Builder = apply {
                checkSettings = true
                locationRequests.add(request)
            }

            fun disableCheckAlways() = apply {
                baseBuilder.disableCheckAlways()
            }

            fun build(): Options {
                return Options(checkSettings, locationRequests, baseBuilder.build())
            }
        }
    }

    override fun onSaveState(state: Bundle) {
        super.onSaveState(state)
    }

    fun checkLocationPermission() {
        checkPermissionsAll(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION))
    }

    private fun beginSettingsCheck(alreadyGranted: Boolean) {

        val client = LocationServices.getSettingsClient(context)
        val task = client.checkLocationSettings(
                LocationSettingsRequest.Builder()
                        .addAllLocationRequests(getOptions().requests)
                        .build()
        )

        task.addOnCompleteListener(context) {
            try {
                val res = it.getResult(ApiException::class.java)
                processLocationSettings(res.locationSettingsStates, alreadyGranted)
            } catch (e: ApiException) {
                when (e.statusCode) {
                    LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
                        val resolvableApiException = e as ResolvableApiException
                        try {
                            doResolveResolvable(resolvableApiException.resolution, requestCodeResolveSettings)
                        } catch (e: SendIntentException) {
                            LoggerFactory.getLogger(TAG).error("Couldn't resolve error: {}", e.stackTraceToString())
                            interceptedCallbacks.onPermissionDenied()
                        }
                    }
                    else -> interceptedCallbacks.onPermissionDenied()
                }
            }
        }
    }

    protected fun processLocationSettings(locationSettingsStates: LocationSettingsStates?, alreadyGranted: Boolean) {
        if (locationSettingsStates == null) {
            // This is happening on mocked location provider
            LoggerFactory.getLogger(TAG).error("Invalid location setting states, blindly grant permission!")
            interceptedCallbacks.onPermissionGranted(alreadyGranted)
            return
        }

        if (locationSettingsStates.isLocationPresent && locationSettingsStates.isLocationUsable) {
            interceptedCallbacks.onPermissionGranted(alreadyGranted)
        } else {
            interceptedCallbacks.onPermissionDenied()
        }
    }

    companion object {
        const val DEFAULT_RC_RESOLVE_SETTINGS = 2
        private const val TAG = "LPermissionDelegate"
        private fun defaultOptions(): Options = Options.Builder().build()
    }
}
